### ISSUE TITLE
TestHelper "broken?" after update to .NET Core 3.1.1

### ISSUE THREAD

**Issue by ErikEJ** [FluentValidation/FluentValidation#1336]
### System Details

- FluentValidation version: 8.6.1 (was using 8.1.2 before)
- Web Framework version: ASP.NET Core 3.1.1

### Issue Description

Runtime validation works as expected, but TestHelper - ShouldHaveValidationErrorFor does not get validation error. 

```
    public class CreateTagParametersValidator : AbstractValidator<CreateTagParameters>
    {
        public CreateTagParametersValidator()
        {
            RuleFor(req => req.OrganizationId).NotEmpty();
            RuleFor(req => req.Request).NotNull();
            RuleFor(req => req.Request.TagName).NotEmpty().When(req => req.Request != null);
            RuleForEach(req => req.Request.Assets).SetValidator(new TagAssetInstanceValidator()).When(req => req.Request != null);
        }
    }

    public class TagAssetInstanceValidator : AbstractValidator<TagAssetInstance>
    {
        public TagAssetInstanceValidator()
        {
            RuleFor(t => t.AssetId).NotEmpty();
        }
    }

    using AutoFixture.Xunit2;

    [Theory, AutoData]
    public void Should_Have_Error_For_Null_AssetId(
            CreateTagParametersValidator validator,
            CreateTagParameters parameters)
        {
           // No validation error here !?
            validator.ShouldHaveValidationErrorFor(parameters, p => p.Request.Assets[0], p => p.Request.Assets[0].AssetId = null);
        }

```

**Comment by JeremySkinner** [comment]
Hi there, please can you provide all the sample code necessary to reproduce the issue - there isn't enough here for me to compile/run. Thanks.

**Comment by ErikEJ** [comment]
Adding missing parts. 


    public interface ICreateTagParameters
    {
        TagCreateDto Request { get; set; }
    }

    public class CreateTagParameters : ICreateTagParameters
    {
        /// <summary>
        /// Tag request body.
        /// </summary>
        [Required]
        [FromBody]
        public TagCreateDto Request { get; set; }
    }

    public class TagCreateDto
    {
        [Required]
        public string TagName { get; set; }
        public string TagDescription { get; set; }
        public string TagColor { get; set; }
        public List<TagAssetInstance> Assets { get; set; }
    }

    public class TagAssetInstance
    {
        [Required]
        public string AssetId { get; set; }
    }


    public static class ValidatorExtensions
    {
        public static void ShouldHaveValidationErrorFor<T, TValue>(this IValidator<T> validator, T data, Expression<Func<T, TValue>> expression, Action<T> prepare)
            where T : class
        {
            prepare(data);

            validator.ShouldHaveValidationErrorFor(expression, data);
        }

        public static void ShouldNotHaveValidationErrorFor<T, TValue>(this IValidator<T> validator, T data, Expression<Func<T, TValue>> expression, Action<T> prepare)
            where T : class
        {
            prepare(data);

            validator.ShouldNotHaveValidationErrorFor(expression, data);
        }
    }

**Comment by ErikEJ** [comment]
Also added missing test attributes to original code

**Comment by JeremySkinner** [comment]
Can you make this a bare minimum case to reproduce, ie strip out AutoFixture and provide a working sample that provides just the input data that's problematic. Thanks.

**Comment by JeremySkinner** [comment]
Also something to check: If you validate manually (ie invoke `Validate` and check the validation results), are the validation errors associated with the properties you expect? Looking at the object design, I woiuld expect the validation failure for the AssetId to be associated with "Request.Assets[0].AssetId" (I'm assuming there's at least 1 asset in the collection, with a null ID), however your assertion is for "Request.Assets[0]", which won't match, so you may need to update your 
assertion to be on "Request.Assets[0].AssetId" (I'll be able to confirm once you have a working sample)

**Comment by ErikEJ** [comment]
Simple repro (sorry for the mad pasting above) - this runs to completion, then update to 8.6.1 and the Should method fails!

.csproj:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="FluentValidation.AspNetCore" Version="8.1.2" />
  </ItemGroup>

</Project>

```

Program.cs:

```
using FluentValidation;
using FluentValidation.TestHelper;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq.Expressions;

namespace ConsoleApp5
{
    class Program
    {
        static void Main(string[] args)
        {
            CreateTagParametersValidator validator = new CreateTagParametersValidator();
            CreateTagParameters parameters = new CreateTagParameters();

            parameters.Request = new TagCreateDto { TagName = "Tag" };
            parameters.Request.Assets = new List<TagAssetInstance> { new TagAssetInstance { AssetId = null } };

            validator.ShouldHaveValidationErrorFor(parameters, p => p.Request.Assets[0], p => p.Request.Assets[0].AssetId = null);
        }

        public class CreateTagParametersValidator : AbstractValidator<CreateTagParameters>
        {
            public CreateTagParametersValidator()
            {
                RuleFor(req => req.Request).NotNull();
                RuleFor(req => req.Request.TagName).NotEmpty().When(req => req.Request != null);
                RuleForEach(req => req.Request.Assets).SetValidator(new TagAssetInstanceValidator()).When(req => req.Request != null);
            }
        }

        public class TagAssetInstanceValidator : AbstractValidator<TagAssetInstance>
        {
            public TagAssetInstanceValidator()
            {
                RuleFor(t => t.AssetId).NotEmpty();
            }
        }

        public interface ICreateTagParameters
        {
            TagCreateDto Request { get; set; }
        }

        public class CreateTagParameters : ICreateTagParameters
        {
            [Required]
            public TagCreateDto Request { get; set; }
        }

        public class TagCreateDto
        {
            [Required]
            public string TagName { get; set; }
            public string TagDescription { get; set; }
            public string TagColor { get; set; }
            public List<TagAssetInstance> Assets { get; set; }
        }

        public class TagAssetInstance
        {
            [Required]
            public string AssetId { get; set; }
        }


    }

    public static class ValidatorExtensions
    {
        public static void ShouldHaveValidationErrorFor<T, TValue>(this IValidator<T> validator, T data, Expression<Func<T, TValue>> expression, Action<T> prepare)
            where T : class
        {
            prepare(data);

            var result = validator.ShouldHaveValidationErrorFor(expression, data);
        }

        public static void ShouldNotHaveValidationErrorFor<T, TValue>(this IValidator<T> validator, T data, Expression<Func<T, TValue>> expression, Action<T> prepare)
            where T : class
        {
            prepare(data);

            validator.ShouldNotHaveValidationErrorFor(expression, data);
        }
    }

}

```

**Comment by JeremySkinner** [comment]
Thanks for the full sample.  Yes this is indeed because of the indexer - you'll need to switch to a string-based comparison for this. Simplest thing to do would be additional overloads of your extension methods that take a string property path:

```csharp
public static void ShouldHaveValidationErrorFor<T>(this IValidator<T> validator, T data, string propertyPath, Action<T> prepare)
	where T : class
{
	prepare(data);

	// Use the TestValidate method directly
	// the standard ShouldHaveValidationErrorFor extension is just a wrapper around this.
	var testResult = validator.TestValidate(data);
	testResult.ShouldHaveValidationErrorFor(propertyPath);
}
```
...which you would then use like this:

```csharp
validator.ShouldHaveValidationErrorFor(parameters, "Request.Assets[0].AssetId", p => p.Request.Assets[0].AssetId = null);

```
We can look at updating the lambda-based overloads to handle indexers properly too - but this is farily compelx as it requires walking the expression tree. I've created an issue to track the development (#1339) but this is probably going to be a fair way off.

**Comment by JeremySkinner** [comment]
Also for future reference this is covered in the "advanced test extensions" section in the documentation: https://docs.fluentvalidation.net/en/latest/testing.html#advanced-test-extensions

**Comment by ErikEJ** [comment]
But it worked in 8.1.2 ??

**Comment by JeremySkinner** [comment]
> But it worked in 8.1.2 ??

I'm afraid not - it was a false-positive. We've never supported expression trees in lambdas like this, but older versions of the testhelper (including 8.1.2) actually *silently failed to run* when an indexer was encountered (leading to a false positive).  This was fixed last year in the 8.5 release, and I also added the string-based overload to work around this limitation.
